frontend/udp/writer.go

```diff
@@ -55,19 +59,26 @@ func WriteAnnounce(w io.Writer, txID []byte, resp *bittorrent.AnnounceResponse,
 }
 
 // WriteScrape encodes a scrape response according to BEP 15.
-func WriteScrape(w io.Writer, txID []byte, resp *bittorrent.ScrapeResponse) {
+func WriteScrape(w io.Writer, txID []byte, infohashes []bittorrent.InfoHash, resp *bittorrent.ScrapeResponse) error {
 	buf := newBuffer()
 
 	writeHeader(buf, txID, scrapeActionID)
 
-	for _, scrape := range resp.Files {
+	// Preserve the ordering.
+	for _, ih := range infohashes {
+		scrape, ok := resp.Files[ih]
+		if !ok {
+			buf.free()
```

Comment:
yea, we had a bug a few years ago because we forgot to release a mutex one place.
i wish defer had slightly different semantics and was basically just a macro. however, i acknowledge it's cost and we should probably avoid it.